### DISCUSSION TITLE
`useTransition` conflicts with React 18 `useTransition` hook

### DISCUSSION THREAD

**Discussion by a7medev** [remix-run/remix#478] [General]
# The Problem

The `useTransition` hook's name conflicts with React 18 `useTransition` hook, I think this needs to be changed to something else so that the Remix API stays stable and doesn't change very often.

**Comment by mdingena** [comment]
You could do `React.useTransition()` instead.

**Reply by matmilbury** [reply]
I don't think it's a matter of variable name conflict.

It's deeper than that.

It's that when you have one word that might refer to two things, the communication becomes confusing.

**Reply by akomm** [reply]
@matmilbury you are 100% right, but can't imagine mdingena meant this as a permanent solution. I guess he just communicated to short.

**Reply by mdingena** [reply]
No, I meant as a permanent solution. Laying claim to function names across libraries is silly. You can deal with this on the consumer side. If you prefer `useTransition` to be a "React thing" then import Remix's version of it with `import { useTransition as useRemixTransition }` and just move on with your life.

**Reply by adebiyial** [reply]
[Yet another JavaScript framework](https://css-tricks.com/yet-another-javascript-framework) is a very good article about naming

**Comment by ryanflorence** [comment]
~~It's unlikely you'll ever interact with `React.useTransition` directly in Remix. We've been using the word "transition" for route transitions long before React grabbed the name. It'll be easy to change later if we need to but we're not concerned about it.~~

[See my other comment](https://github.com/remix-run/remix/discussions/478#discussioncomment-2555551)

**Reply by kag359six** [reply]
This is a bit off topic, but isn't the `React.useTransition` API still useful when trying to prevent the browser from freezing up when doing expensive state updates? For example, changing a filter for a data intensive chart, or updating a large form when using a library like Formik that re-renders the entire form on state change. 

Unless I'm misunderstanding that specific React API, or perhaps Remix has it's own way around those scenarios.

**Comment by ryanflorence** [comment]
FYI: As we move the remix data APIs to react router, we've renamed the hook `useNavigation`

```jsx
let navigation = useNavigation();
navigation.state
navigation.formData
// etc.
```

When we re-implement Remix on top of the next version of React Router, we'll support both hooks in parallel (`useTransition` and `useNavigation`) and add a warning in `useTransition` to switch to `useNavigation`.

Then in v2 we'll remove `useTransition`.

I like `useNavigation` better, too, so I'm pretty happy about this.

(I still think React's should be called `useUpdate`, and then animation libs could keep `useTransition`. But it's cool 💙🧊💙🧊💙🧊💙)

**Reply by mikeybinns** [reply]
Currently Remix doesn't export the `Transition` type so we have to do `ReturnType<typeof useTransition>`, Is this because of this name conflict or is this just an oversight?
I feel like it should be exported because the Fetcher type is exported from that package, but perhaps I'm missing something?
